bundled/tool/ruff_server.py

```diff
@@ -0,0 +1,89 @@
+import sys
+import sysconfig
+from pathlib import Path
+
+RUFF_EXE = "ruff.exe" if sys.platform == "win32" else "ruff"
+
+BUNDLE_DIR = Path(__file__).parent.parent
+
+
+def update_sys_path(path_to_add: str) -> None:
+    """Add given path to `sys.path`."""
+    if os.path.isdir(path_to_add):
+        # The `site` module adds the directory at the end, if not yet present; we want
+        # it to be at the beginning, so that it takes precedence over any other
+        # installed versions.
+        sys.path.insert(0, path_to_add)
+
+        # Allow development versions of libraries to be imported.
+        site.addsitedir(path_to_add)
+
+
+if __name__ == "__main__":
+    # Ensure that we can import bundled libraries like `packaging`
+    update_sys_path(os.fspath(BUNDLE_DIR / "libs"))
+
+
+from packaging.specifiers import SpecifierSet
+from packaging.version import Version
+
+# This is the first release that included `ruff server`.
+# The minimum version may change in the future.
+RUFF_VERSION_REQUIREMENT = SpecifierSet(">=0.3.3")
+# These versions have major bugs or broken integration, and should be avoided.
+FORBIDDEN_RUFF_VERSIONS = [Version("0.3.4")]
+
+
+def executable_version(executable: str) -> Version:
+    """Return the version of the executable at the given path."""
+    output = subprocess.check_output([executable, "--version"]).decode().strip()
+    version = output.replace("ruff ", "")
+    return Version(version)
+
+
+def check_compatibility(
+    executable: str,
+    requirement: SpecifierSet,
+    forbidden_requirements: SpecifierSet,
+) -> None:
+    """Check the executable for compatibility against various version specifiers."""
+    version = executable_version(executable)
+    if not requirement.contains(version, prereleases=True):
+        message = f"Ruff {requirement} required, but found {version} at {executable}"
+        raise RuntimeError(message)
+    for forbidden in forbidden_requirements:
+        if version == forbidden:
+            message = (
+                f"Tried to use Ruff version {version} at {executable}, which is not allowed.\n"
+                "This version of the server has incompatibilities and/or integration-breaking bugs.\n"
+                "Please upgrade to the latest version and try again."
+            )
+            raise RuntimeError(message)
+
+
+def find_ruff_bin(fallback: Path) -> Path:
+    """Return the ruff binary path."""
+    path = Path(sysconfig.get_path("scripts")) / RUFF_EXE
+    if path.is_file():
+        return path
+
+    path = shutil.which("ruff")
+    if path:
+        return path
+
+    return fallback
+
+
+if __name__ == "__main__":
+    ruff = os.fsdecode(
+        find_ruff_bin(
+            Path(BUNDLE_DIR / "libs" / "bin" / RUFF_EXE),
+        ),
+    )
+    check_compatibility(ruff, RUFF_VERSION_REQUIREMENT, FORBIDDEN_RUFF_VERSIONS)
+    completed_process = subprocess.run([ruff, *sys.argv[1:]], check=False)
+    sys.exit(completed_process.returncode)
```

Comment:
In Ruff, we do:

```python
if __name__ == "__main__":
    ruff = os.fsdecode(find_ruff_bin())
    if sys.platform == "win32":
        import subprocess

        completed_process = subprocess.run([ruff, *sys.argv[1:]])
        sys.exit(completed_process.returncode)
    else:
        os.execvp(ruff, [ruff, *sys.argv[1:]])
```

Can we use that same logic here?

Reply:
This was actually how the code looked when I first ported it over - I can revert it, but I would like to understand why we can't simplify this to a single `subprocess.run` call.

(Ruff threw [S606](https://docs.astral.sh/ruff/rules/start-process-with-no-shell/) for our usage of `os.execvp`, recommending that we use `subprocess` instead)

Reply:
I'd suggest looking back at the Git history on that file. I don't know the answer off-hand but I suspect there's more context in the history.

Reply:
It looks like this was added to improve responsiveness on startup: https://github.com/astral-sh/ruff/commit/158367bf913f5a5df2c36f0d160ac35f96b5364c.

Which honestly doesn't seem that necessary in this case, since we aren't repeatedly starting up the server. I'd prefer to keep this as a simple `subprocess.run` invocation.